js/pathfinding.js

```diff
@@ -55,27 +53,51 @@ export function findPath(from, to, token, previousWaypoints) {
 	}
 }
 
-function getNode(pos, token, initialize=true) {
-	if (!cachedNodes)
-		cachedNodes = new Array(gridHeight);
-	if (!cachedNodes[pos.y])
-		cachedNodes[pos.y] = new Array(gridWidth);
-	if (!cachedNodes[pos.y][pos.x]) {
-		cachedNodes[pos.y][pos.x] = pos;
+/**
+ * Build a cache ID based on the current token's data and then retrieve the cache to use from that
+ */
+function getCachedNodes(token) {
+	const cacheData = {};
+
+	// Different-sized tokens snap to different points on the grid,
+	// so they might follow a different path to other tokens
+	cacheData.tokenWidth = token.width;
+	cacheData.tokenHeight = token.height;
+
+	// If levels is enabled, the token's elevation can affect which walls
+	// they need to worry about
+	if (game.modules.get("levels")?.active) {
+		cacheData.elevation = token.data.elevation;
```

Comment:
Will we now create a new cache for every single elevation? I think that's a little excessive. We maybe need an upper limit for existing caches (or some smarter detection of which elevation ranges are actually relevant).

Reply:
Yes, if Levels is enabled we'll create a cache for every token size/elevation combination. With a fair chunk more work, we could see if the new cache would be functionally equivalent to an existing one, but that would require the caches be complete.